### PR TITLE
SOLR-14316 Remove unchecked type conversion warning in JavaBinCodec's readMapEntry's equals() method

### PR DESCRIPTION
**Pull request by aroopganguly**
# Description

there was an unchecked type conversion warning in JavaBinCodec's readMapEntry's equals() method

# Solution
fixed an unchecked type conversion warning in JavaBinCodec's readMapEntry's equals() method

# Tests

added tests per @noblepaul 's comments

- please do note that no tests existed for this api prior to this so I had to think how to set a baseline here
- however now we have a a test case for this api oh JavaBinCodec, with decent boundary value code coverage
- I tested for backward compatibility between the code change by running the same boundary value analysis of tests, with the new and old api implementation and the test results are the same

Some Observations
- so from these tests, one can surmise that the JavaBinCodec.readMapEntry() method's current behavior is to create/expect new Objects to be created from binary streams.
- What the tests do reveal is that its possible to use this api to read text streams too where the api does not generate object but uses primitive string comparisons.
- Which is fine as I think this api was and is meant to be used for binary file renditions only.
- however one may think that this should be more formally enforced ?

# Checklist

Please review the following and check all that apply:

- [x] I have reviewed the guidelines for [How to Contribute](https://wiki.apache.org/solr/HowToContribute) and my code conforms to the standards described there to the best of my ability.
- [x] I have created a Jira issue and added the issue ID to my pull request title.
- [x] I have given Solr maintainers [access](https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to contribute to my PR branch. (optional but recommended)
- [x] I have developed this patch against the `master` branch.
- [x] I have run `ant precommit` and the appropriate test suite.
- [x] I have run `gradlew precommit` and the appropriate test suite.
- [x] I have added tests for my changes.
- [ ] I have added documentation for the [Ref Guide](https://github.com/apache/lucene-solr/tree/master/solr/solr-ref-guide) (for Solr changes only).

### CODE REVIEW

**Comment by noblepaul** [comment]
LGTM

a test can help

**Comment by aroopganguly** [comment]
@noblepaul tests added and some points to ponder upon as well in the Tests section in the main section of the PR above.

**Comment by anshumg** [comment]
LGTM. Thanks for adding the test.

Can you please also add a CHANGELOG entry?

**Comment by aroopganguly** [comment]
Thanks @anshumg I have added a changelog entry as advised.

File: solr/solrj/src/test/org/apache/solr/common/util/TestJavaBinCodec.java [hunk 1]
```diff
@@ -287,6 +289,45 @@ public void testAllTypes() throws IOException {
...
+  @Test
+  public void testReadMapEntry() throws IOException {
+    Map.Entry<Object, Object> extryFromTextDoc1 = getMapFromJavaBinCodec(SOLRJ_DOCS_1);
+    Map.Entry<Object, Object> extryFromTextDoc1_clone = getMapFromJavaBinCodec(SOLRJ_DOCS_1);
+
+    Map.Entry<Object, Object> extryFromTextDoc2 = getMapFromJavaBinCodec(SOLRJ_DOCS_2);
+    Map.Entry<Object, Object> extryFromTextDoc2_clone = getMapFromJavaBinCodec(SOLRJ_DOCS_2);
+
+    // exactly same document read twice should have same content
+    assert (extryFromTextDoc1.equals(extryFromTextDoc1_clone));
```
Comment (tflobbe) [inline solr/solrj/src/test/org/apache/solr/common/util/TestJavaBinCodec.java]: While this would work because we force to enable assertions while running tests, we usually use JUnit's assertions, such as `assertEquals("", obj1, obj2);`

File: solr/solrj/src/test/org/apache/solr/common/util/TestJavaBinCodec.java [hunk 2]
```diff
@@ -287,6 +289,45 @@ public void testAllTypes() throws IOException {
...
+    // doc1 and doc2 are 2 text files with different content on line 1
+    assert (!extryFromTextDoc2.equals(extryFromTextDoc1));
+    // exactly same document read twice should have same content
+    assert (extryFromTextDoc2.equals(extryFromTextDoc2_clone));
+
+    // now lets look at binary files
```
Comment (tflobbe) [inline solr/solrj/src/test/org/apache/solr/common/util/TestJavaBinCodec.java]: suggestion, maybe take this to a different method?

File: solr/solrj/src/test/org/apache/solr/common/util/TestJavaBinCodec.java [hunk 3]
```diff
@@ -287,6 +289,45 @@ public void testAllTypes() throws IOException {
...
+    Map.Entry<Object, Object> entryFromBinFileA = getMapFromJavaBinCodec(SOLRJ_JAVABIN_BACKCOMPAT_BIN);
+    Map.Entry<Object, Object> entryFromBinFileA_clone = getMapFromJavaBinCodec(SOLRJ_JAVABIN_BACKCOMPAT_BIN);
+
+    // as the equals assertion is based on the hashcode, even the same exact bin file
+    // will produce different Objects and hence different hashcodes
```
Comment (tflobbe) [inline solr/solrj/src/test/org/apache/solr/common/util/TestJavaBinCodec.java]: this is interesting. what's the type of the objects in the map for this case?

File: solr/solrj/src/test/org/apache/solr/common/util/TestJavaBinCodec.java (hunk 1 above)
Comment (aroopganguly) [reply solr/solrj/src/test/org/apache/solr/common/util/TestJavaBinCodec.java]: sure !

File: solr/solrj/src/test/org/apache/solr/common/util/TestJavaBinCodec.java (hunk 2 above)
Comment (aroopganguly) [reply solr/solrj/src/test/org/apache/solr/common/util/TestJavaBinCodec.java]: sure

**Comment by aroopganguly** [comment]
closing this PR in favor of https://github.com/apache/lucene-solr/pull/1344
